### ISSUE TITLE
Support clang's -frewrite-includes flag with CCACHE_NOCPP2 by sending -D defines to both preprocess and compile

### ISSUE THREAD

**Issue by RedBeard0531** [ccache/ccache#185]
Currently we need to use CCACHE_CPP2 when building with clang due to -Werror tripping differently on preprocessed output. clang has a -frewrite-includes flag that causes it to only rewrite #include lines without doing the full preprocessing to handle just this use case (it is especially important with icecream/distcc where the equivalent of CCACHE_CPP2 isn't an option). Unfortunately this doesn't work with CCACHE_NOCPP2 because the -D flags are stripped out when running the compiler. Additionally, it requires passing -xc rather than -xcpp-output since it hasn't been fully preprocessed yet. It may be worth always passing the -D flags unless it is actually harmful on some compilers.

gcc has a similar -fdirectives-only flag, but it includes all -D flags from the command-line in the preprocessed output, so it works correctly with ccache today.

**Comment by afbjorklund** [comment]
What about pump mode, for distcc ?

**Comment by RedBeard0531** [comment]
I think pump mode philosophically conflicts with ccache. The point of pump mode is for when the load of preprocessing locally is too high, and you want to offload it to remote nodes. ccache requires local preprocessing, although by default it currently tells the compiler to do the preprocessing again from scratch. If you can't preprocess locally, you can't really use ccache. My goal is to make it possible to use ccache with icecream (which doesn't have a pump mode), while only preprocessing locally once rather than twice.

**Comment by afbjorklund** [comment]
Well, pump happens to work for cpp2 as well by not sending / requiring .ii but sending the .cc

With the latest clang and c++ changes, preprocessing is getting obsolete... It's useful as an intermediate step but we don't save it anymore. Similar to the assembly, I suppose.

**Comment by RedBeard0531** [comment]
Sure, but ccache is still doing a full preprocess pass locally before invoking distcc in pump mode where another full preprocess pass will be done. My goal is to only do a single pre-process pass, possibly split into one pass for handling #includes and one to do the real preprocessing. The problem is that ccache running the preprocessor uses a significant amount of CPU cycles and makes my builds ~50% slower.

My goal could be served either by having ccache do the #include-only preprocessing (the subject of this ticket), or by having ccache do no preprocessing at all and only rely on direct-mode hashing. It looks like ccache has a way to disable direct mode, but no way to disable non-direct mode using cpp. I assumed that was an intentional design decision, but if that is more likely to happen, I can open a separate ticket for that.

**Comment by afbjorklund** [comment]
Hmm, wonder if I didn't do something like this earlier when I misunderstood how `-f-directives-only` work ?

That "avoid non-direct" option might be a good idea, it's probably an accident waiting to happen otherwise.

**Comment by afbjorklund** [comment]
Back to your original statement, I don't actually think there is any significant difference in how gcc handles `-fdirectives-only` and how clang handles `-frewrite-includes`. It seems that *both* compilers choke on the half-preprocessed input, unless you pass the same option again also to the compiler (enabling macros).

But it does require some code, since the manifest relies on the preprocessor in order to find the files (so can't really disable it) and like you say the macros need some special handling if they are to work with (the new limited) cpp1. Probably one could use the same for both compilers, but that remains to be seen...

**Comment by afbjorklund** [comment]
@RedBeard0531 : Can you try https://github.com/ccache/ccache/compare/master...afbjorklund:cpp1 ?

It should work OK for both gcc/clang, and if it does I can make it into a regular Pull Request later on.
(there is a test case that seemed reasonable, but haven't yet tried it with distcc or anything though)

Note that GCC uses `-fpreprocessed`, and doesn't like to be told what to do with `-x` or anything...
So we only tell Clang the original language, just so that it doesn't forget to parse the macros again.

**Comment by RedBeard0531** [comment]
Thanks, that looks like it should do the trick. I'll give it a shot sometime this week.

> Back to your original statement, I don't actually think there is any significant difference in how gcc handles -fdirectives-only and how clang handles -frewrite-includes. It seems that both compilers choke on the half-preprocessed input, unless you pass the same option again also to the compiler (enabling macros).

That is partially true. IMO GCC has the better implementation here. It processes #if blocks locally which generally results in a much smaller file. Additionally it converts command-line -D arguments to #defines in the .ii file so they don't need to be passed again to the compiler (although it shouldn't be harmful to do so). However, unlike clang, it requires that the -fdirectives-only flag be passed to both passes, while clang's -frewrite-includes only needs to be passed to the cpp phase. 

That said, your patch treating them the same seems like it should be fine.

**Comment by afbjorklund** [comment]
Yeah, there's probably more that _could_ be done here by splitting the `cpp_args` up further...
e.g. things only setting macros (like `-D`) and things only setting preprocessor paths (like `-I`)

And one could also improve this a little bit, in terms of expressiveness and avoid redundancy,
by having gcc add `-fpreprocessed` and avoiding sending `-frewrite-includes` to compiler.

----
BTW; In addition to the "\<command-line\>" mentioned my test case also gets a ton of _other_ stuff,
from the "\<built-in>\" and from "/usr/include/stdc-predef.h" - but I guess it's small* when compared.

e.g. about 300 lines, which is a bit big for a C program but nothing when compared with a C++ one.
Pretty much anything from Boost or the Standard Library pushes the line count up into the 1000's ?

**Comment by RedBeard0531** [comment]
On Mon, Jun 26, 2017 at 4:32 PM, Anders Björklund <notifications@github.com>
wrote:

> Yeah, there's probably more that *could* be done here by splitting the
> cpp_args up further...
> e.g. things only setting macros (like -D) and things only setting
> preprocessor paths (like -I)
>
> And one could also improve this a little bit, in terms of expressiveness
> and avoid redundancy,
> by having gcc add -fpreprocessed and avoiding sending -frewrite-includes
> to compiler.
>
Meh. Probably best to do the simplest thing that works. Seems least likely
to need futzing with to keep working in the future.

> BTW; In addition to the "<command-line>" mentioned my test case also gets
> a ton of *other* stuff,
> from the "<built-in>" and from "/usr/include/stdc-predef.h" - but I guess
> it's small* when compared.
>
> e.g. about 300 lines, which is a bit big for a C program but nothing when
> compared with a C++ one.
> Pretty much anything from Boost or the Standard Library pushes the line
> count up into the 1000's ?
>
1000's of lines is nothing. With the .cpp I was testing I think the
preprocessed size was around 5MB with clang++ and 1-2MB with g++. It's been
a while since I looked so I don't remember the exact numbers, but those are
the right orders of magnitude.

>

**Comment by afbjorklund** [comment]
> Meh. Probably best to do the simplest thing that works. Seems least likely
to need futzing with to keep working in the future.

Well, I already wrote it ;-)

> 1000's of lines is nothing.

Nothing for a "real" program, but definitely is something just for an "empty" one.

I guess modules are the real solution for big header libraries like these, but alas
it didn't make it into the standard proper (and will also be problems for distcc...)

**Comment by afbjorklund** [comment]
BTW; If you are interested in this you might want to look into https://github.com/distcc/distcc/pull/168 (and friends...)
For it to work, there's a bunch of things that need to be fixed in distcc. Maybe "icecream" did ?

**Comment by RedBeard0531** [comment]
Sorry for the delay, I just got a chance to try your patch. It works great! 

I compared CCACHE_RECACHE=1 builds both with and without -fdirectives-only/-frewrite-includes and they make builds take 30-50% less time. They are now very close to CCACHE_DISABLE=1 builds. 

I don't think that distcc issue affects icecream, since I don't think it limits parallelism of cpp runs (unlike local compile jobs). Also this patch would be a nice work around for that anyway since it preprocesses before distcc gets a chance to lock.
